### PR TITLE
Allow data submission for non fxos-devices

### PR DESCRIPTION
**Pull request by johngian**
- Change devices models/forms to facilitate this
- Add ui to select between fxos/freeform model/manufacturer

### CODE REVIEW

File: marketpulse/devices/forms.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    def __init__(self, *args, **kwargs):
+        required_device = kwargs.pop('device', False)
+        super(DeviceForm, self).__init__(*args, **kwargs)
+        self.fields['model'].required = required_device
+
+    def save(self, *args, **kwargs):
+        query = {
+            "model__icontains": self.cleaned_data['model'],
+            "manufacturer__icontains": self.cleaned_data['manufacturer']
```
Comment (akatsoulas) [inline marketpulse/devices/forms.py]: Probably makes sense to use `iexact` here.

File: marketpulse/devices/models.py [hunk 2]
```diff
@@ -5,10 +5,12 @@ class Device(models.Model):
     """Model for FfxOS devices data."""
 
     model = models.CharField(max_length=120)
-    manufacturer = models.CharField(max_length=120)
+    manufacturer = models.CharField(max_length=120, blank=True, null=True)
```
Comment (akatsoulas) [inline marketpulse/devices/models.py]: Why did you add here both `blank` and `null`. Shouldn't a device always have a manufacturer either is a FxOS device or not?

**Comment by johngian** [comment]
@akatsoulas PR Updated

Tested it locally:
- Add/Edit new contribution with fxos device
- Add/Edit new contribution with non fxos device
- Form validation for fxos
- Form validation for non-fxos
- UI (initialize form depending on fxos-toggle. show/hide)

File: marketpulse/devices/forms.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            "model__iexact": self.cleaned_data['model'],
+            "manufacturer__iexact": self.cleaned_data['manufacturer']
+        }
+
+        qs = Device.objects.filter(**query)
+        if qs.exists():
+            return qs[0]
+
+        device = super(DeviceForm, self).save(*args, **kwargs)
+        device.is_fxos = False
```
Comment (akatsoulas) [inline marketpulse/devices/forms.py]: Is this needed now that you are using a modelform?

File: marketpulse/devices/forms.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        manufacturer = cleaned_data.get('manufacturer')
+        is_fxos = cleaned_data.get('is_fxos')
+        cleaned_data['is_fxos'] = bool(is_fxos)
+
+        if not is_fxos:
+            msg = 'This field is required'
+            if not model:
+                self._errors['model'] = self.error_class([msg])
+            if not manufacturer:
+                self._errors['manufacturer'] = self.error_class([msg])
```
Comment (akatsoulas) [inline marketpulse/devices/forms.py:41]: Really minor but you can you use `any` to check if either of the two fields is missing. Probably this will make the clean method much shorter. Feel free to ignore.

Comment (johngian) [inline marketpulse/devices/forms.py:41]: I know but you also need to determine which field has the error.

**Comment by johngian** [comment]
@akatsoulas PR updated

File: marketpulse/devices/forms.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+            "model__iexact": self.cleaned_data['model'],
+            "manufacturer__iexact": self.cleaned_data['manufacturer']
+        }
+
+        qs = Device.objects.filter(**query)
+        if qs.exists():
+            return qs[0]
+
+        device = super(DeviceForm, self).save(*args, **kwargs)
+        device.save()
```
Comment (akatsoulas) [inline marketpulse/devices/forms.py]: you can just use `return super(DeviceForm, self).save(*args, **kwargs)`

Comment (johngian) [inline marketpulse/devices/forms.py]: True dat

File: marketpulse/devices/models.py [hunk 6]
```diff
@@ -4,11 +4,13 @@
...
+    model = models.CharField(max_length=120, blank=True)
+    manufacturer = models.CharField(max_length=120, blank=True)
+    is_fxos = models.BooleanField(default=True)
 
     def __unicode__(self):
         return u'{0}, {1}'.format(self.manufacturer, self.model)
 
     class Meta:
         ordering = ['manufacturer', 'model']
+        unique_together = ('model', 'manufacturer')
```
Comment (akatsoulas) [inline marketpulse/devices/models.py]: nit: Please append a `,`

File: marketpulse/devices/models.py [hunk 7]
```diff
@@ -4,11 +4,13 @@
 class Device(models.Model):
     """Model for FfxOS devices data."""
 
-    model = models.CharField(max_length=120)
-    manufacturer = models.CharField(max_length=120)
+    model = models.CharField(max_length=120, blank=True)
+    manufacturer = models.CharField(max_length=120, blank=True)
```
Comment (akatsoulas) [inline marketpulse/devices/models.py]: Please add a default value for `model` and `manufacturer`.

File: marketpulse/main/static/js/fxosprice_new.js [hunk 8]
```diff
@@ -66,4 +78,12 @@ $(document).ready(function() {
 
     // Show/Hide price formset remove button
     $('.price-remove').ready(displayRemovePriceBtn);
+
+    // Initial other-device form on load
+    var is_fxos = $('#id_is_fxos').attr('checked');
```
Comment (akatsoulas) [inline marketpulse/main/static/js/fxosprice_new.js]: You could omit the `is_fxos` var.

File: marketpulse/main/static/js/fxosprice_new.js [hunk 9]
```diff
@@ -57,6 +57,18 @@ $(document).ready(function() {
...
 
+    $('.is-fxos input').change(function() {
+      if (this.checked) {
+        $('.other-device').hide();
+        $('.fxos-device').show();
+        $('.other-device input').val('');
+      } else {
+        $('.other-device').show();
+        $('.fxos-device').hide();
+        $('.fxos-device select').val('');
```
Comment (akatsoulas) [inline marketpulse/main/static/js/fxosprice_new.js]: If you add a default value to the model, probably you don't need this - haven't tested yet though

File: marketpulse/main/static/js/fxosprice_new.js [hunk 10]
```diff
@@ -57,6 +57,18 @@ $(document).ready(function() {
...
+      }
+    });
+
```
Comment (comzeradd) [inline marketpulse/main/static/js/fxosprice_new.js:69]: Please use 4 spaces for js indentation.

File: marketpulse/main/templates/fxosprice_new.html [hunk 11]
```diff
@@ -63,7 +63,20 @@ <h2 class="title">Firefox OS prices</h2>
 
         <fieldset>
           <legend>Device</legend>
-          <div class="form-group">
+          <div class="switch is-fxos">
+            <label >Is it Firefox OS?</label>
```
Comment (akatsoulas) [inline marketpulse/main/templates/fxosprice_new.html]: There is an extra space here.

Comment (akatsoulas) [inline marketpulse/main/templates/fxosprice_new.html]: What about `Is this a Firefox OS device?` ?

**Comment by akatsoulas** [comment]
Please add docstrings to any new methods that don't have any.

File: marketpulse/main/templates/fxosprice_new.html (hunk 11 above)
Comment (akatsoulas) [inline marketpulse/main/templates/fxosprice_new.html]: Also please add a `for="id_is_fxos"`.

**Comment by johngian** [comment]
@akatsoulas PR updated

**Comment by akatsoulas** [comment]
Nicely done! I don't know if it makes any sense to harden the validator in the free form device. eg allow only [A-Za-z] chars in manufacturer section. I leave this up to you. r+(wc?) :rocket:
